src/core/vectortile/qgsvectortilemvtdecoder.cpp

```diff
@@ -0,0 +1,318 @@
+{
+  if ( !tile.ParseFromArray( rawTileData.constData(), rawTileData.count() ) )
+    return false;
+
+  mTileID = tileID;
+
+  mLayerNameToIndex.clear();
+  for ( int layerNum = 0; layerNum < tile.layers_size(); layerNum++ )
+  {
+    const ::vector_tile::Tile_Layer &layer = tile.layers( layerNum );
+    QString layerName = layer.name().c_str();
+    mLayerNameToIndex[layerName] = layerNum;
+  }
+  return true;
+}
+
+QStringList QgsVectorTileMVTDecoder::layers() const
+{
+  QStringList layerNames;
+  for ( int layerNum = 0; layerNum < tile.layers_size(); layerNum++ )
+  {
+    const ::vector_tile::Tile_Layer &layer = tile.layers( layerNum );
+    QString layerName = layer.name().c_str();
+    layerNames << layerName;
+  }
+  return layerNames;
+}
+
+QStringList QgsVectorTileMVTDecoder::layerFieldNames( const QString &layerName ) const
+{
+  if ( !mLayerNameToIndex.contains( layerName ) )
+    return QStringList();
+
+  const ::vector_tile::Tile_Layer &layer = tile.layers( mLayerNameToIndex[layerName] );
+  QStringList fieldNames;
+  for ( int i = 0; i < layer.keys_size(); ++i )
+  {
+    QString fieldName = layer.keys( i ).c_str();
+    fieldNames << fieldName;
+  }
+  return fieldNames;
+}
+
+QgsVectorTileFeatures QgsVectorTileMVTDecoder::layerFeatures( const QMap<QString, QgsFields> &perLayerFields, const QgsCoordinateTransform &ct ) const
+{
+  QgsVectorTileFeatures features;
+
+  int numTiles = static_cast<int>( pow( 2, mTileID.zoomLevel() ) ); // assuming we won't ever go over 30 zoom levels
+  double z0xMin = -20037508.3427892, z0yMin = -20037508.3427892;
+  double z0xMax =  20037508.3427892, z0yMax =  20037508.3427892;
+  double tileDX = ( z0xMax - z0xMin ) / numTiles;
+  double tileDY = ( z0yMax - z0yMin ) / numTiles;
+  double tileXMin = z0xMin + mTileID.column() * tileDX;
+  double tileYMax = z0yMax - mTileID.row() * tileDY;
+
+  for ( int layerNum = 0; layerNum < tile.layers_size(); layerNum++ )
+  {
+    const ::vector_tile::Tile_Layer &layer = tile.layers( layerNum );
+
+    QString layerName = layer.name().c_str();
+    QVector<QgsFeature> layerFeatures;
+    QgsFields layerFields = perLayerFields[layerName];
+
+    // figure out how field indexes in MVT encoding map to field indexes in QgsFields (we may not use all available fields)
+    QHash<int, int> tagKeyIndexToFieldIndex;
+    for ( int i = 0; i < layer.keys_size(); ++i )
+    {
+      int fieldIndex = layerFields.indexOf( layer.keys( i ).c_str() );
+      if ( fieldIndex != -1 )
+        tagKeyIndexToFieldIndex.insert( i, fieldIndex );
+    }
+
+    // go through features of a layer
+    for ( int featureNum = 0; featureNum < layer.features_size(); featureNum++ )
+    {
+      const ::vector_tile::Tile_Feature &feature = layer.features( featureNum );
+
+      QgsFeature f( layerFields, static_cast<QgsFeatureId>( feature.id() ) );
+
+      //
+      // parse attributes
+      //
+
+      for ( int tagNum = 0; tagNum < feature.tags_size(); tagNum += 2 )
```

Comment:
```suggestion
      for ( int tagNum = 0; tagNum + 1 < feature.tags_size(); tagNum += 2 )
```
would be more prudent in case of corrupted tile with odd number of values in tags[]